resolwe/elastic/indices.py

```diff
@@ -0,0 +1,122 @@
+
+
+class BaseIndex(object):
+    """Base index processor to be used in apps.
+
+    To work properly, subclass of this class must ovverride following methods
+    and attributes:
+
+      * object_type - class to which object must belong to be processed
+      * process_object - method that is called when object is created/updated
+
+    """
+
+    obj = None
+    object_type = None
+    mapping = {}
+    document_class = None
+    _preprocessed = False
+
+    def __init__(self, obj):
+        self.obj = obj
+
+    def filter(self):
+        pass
+
+    def preprocess_object(self):
+        pass
+
+    def generate_id(self):
+        object_type = self.object_type.__name__.lower()
+        index_name = self.document_class()._get_index()
+        return '{}_{}_{}'.format(object_type, index_name, self.obj.id)
+
+    def process_object(self):
+        if not self.object_type:
+            raise RuntimeError('`object_type` must be defined')
+
+        if not isinstance(self.obj, self.object_type):
+            return
+
+        if self.filter() is False:
+            return
+
+        if not self._preprocessed:
+            self.preprocess_object()
+            self._preprocessed = True
+
+        if not self.document_class:
+            raise RuntimeError('`document_class` must be defined')
+
+        document = self.document_class(meta={'id': self.generate_id()})
+
+        for field in document._doc_type.mapping:
+            if field in ['users_with_permissions', 'groups_with_permissions']:
+                continue  # this fields are handled separately
+
+            get_value_function = getattr(self, 'get_{}_value'.format(field), None)
+            if get_value_function:
+                setattr(document, field, get_value_function())
+                continue
+
+            if field in self.mapping:
+                if callable(self.mapping[field]):
+                    setattr(document, field, self.mapping[field]())
+                    continue
+
+                if isinstance(self.mapping[field], six.string_types):
+                    object_attr = dict_dot(self.obj, self.mapping[field], default=self.mapping[field])
+                else:
+                    object_attr = self.mapping[field]
+                if callable(object_attr):
+                    setattr(document, field, object_attr())
+                else:
+                    setattr(document, field, object_attr)
+                continue
+
+            object_value = dict_dot(self.obj, field, default=NoValue())
+
+            if isinstance(object_value, NoValue):
+                raise AttributeError('Cannot determine mapping for field {}'.format(field))
+
+            setattr(document, field, object_value)
+
+        permissions = self.get_permissions()
+        document.users_with_permissions = permissions['users']
+        document.groups_with_permissions = permissions['groups']
+
+        document.save()
+
+    def get_permissions(self):
+        return {
```

Comment:
At the moment each object is processed individually, so I don't see easy option to optimise this. But this is definitely something that we have to work on in the future. I will add a comment about this.